constraint-solver/src/solver.rs

```diff
@@ -88,83 +88,42 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display + Debug> Solver<T, V>
         }
     }
 
-    fn apply_assignment(&mut self, variable: V, expr: SymbolicExpression<T, V>) -> bool {
-        let SymbolicExpression::Concrete(value) = expr else {
-            panic!("Unexpected non-concrete assignment: {variable} = {expr}");
-        };
-
-        let entry = self.variable_states.entry(variable.clone()).or_default();
-        let variable_update = match &entry {
-            // We already know the expression
-            VariableState::Known(existing) => {
-                assert_eq!(
-                    *existing, value,
-                    "Inconsistent assignment for {variable}: {existing} != {value}"
-                );
-                return false;
-            }
-
-            // An unknown variable became known.
-            VariableState::Unknown(..) => {
-                log::trace!("{variable} = {value}");
-
-                *entry = VariableState::Known(value);
-
-                // The borrow checker won't let us call `update_constraints` here, so we
-                // just return the update.
-                Some(VariableUpdate {
-                    variable,
-                    known: true,
-                    range_constraint: RangeConstraint::from_value(value),
-                })
-            }
-        };
-
-        if let Some(variable_update) = variable_update {
-            self.update_constraints(&variable_update);
-            true
-        } else {
-            false
-        }
+    fn apply_assignment(&mut self, variable: &V, expr: &SymbolicExpression<T, V>) -> bool {
+        self.apply_range_constraint_update(variable, expr.range_constraint())
     }
 
     fn apply_range_constraint_update(
         &mut self,
-        variable: V,
+        variable: &V,
         range_constraint: RangeConstraint<T>,
     ) -> bool {
-        let entry = self.variable_states.entry(variable.clone()).or_default();
-        let existing_range_constraint = entry.range_constraint();
-        let updated_constraint = existing_range_constraint.conjunction(&range_constraint);
-
-        if existing_range_constraint == updated_constraint {
-            // Already knew the constraint, no progress
-            return false;
-        }
-
-        log::trace!("({variable}: {updated_constraint})");
-
-        if let VariableState::Unknown(existing) = entry {
-            *existing = updated_constraint.clone();
+        if self.range_constraints.update(variable, &range_constraint) {
+            // The range constraint was updated.
+            log::trace!("({variable}: {range_constraint})");
+
+            let new_rc = self.range_constraints.get(variable);
+            if new_rc.try_to_single_value().is_some() {
+                // The variable is now known.
+                self.update_constraints(&VariableUpdate {
+                    variable: variable.clone(),
+                    known: true,
+                    range_constraint: new_rc.clone(),
+                });
```

Comment:
Right, and we don't have to update the range constraint in the unknown case, because then it's not stored in the the QSE anyway.